Split liquid_viscosity to liquid_viscosity and move_resistance

Wuzzy2 opened the pull request:
Fixes #10736. Fixes #1822.

The goal of this PR is to completely decouple liquid flowing/spreading from liquid player movement physics.

First, this splits off the `liquid_viscosity` feature of nodes into two:

* `liquid_viscosity`: 0-7. This will continue to affect the liquid flowing speed as before, but no longer the movement speed of players.
* `move_resistance`: 0-7. This will slow down players as if they are in a liquid. 0 means no slowdown at all. If this field was not set, it will be equal to `liquid_viscosity` (for backwards-compability).

`move_resistance` also works for non-liquids (except `climbable` nodes). For non-liquids, the physics are very similar to liquids, so the slowdown applies in all directions (including up/down). The biggest difference is that you can't "swim" up or down. The "fluidity" player settings are ignored if the node is not a liquid, in this case, two hardcoded values are used for the physics (see code for details).

To keep simple, the slowdown for `move_resistance` is the same like it was for `liquid_viscosity` before. There is one exception, however: If `move_resistance` is 0, there's no slowdown. Before, the value 0 for old `liquid_viscosity` caused a very subtle bit of slowdown (barely noticable, you have to measure it to see the difference). This mini-slowdown for value 0 was removed.

Second, `liquidtype` has been decoupled to split off player movement physics from the liquid flow mechanic (these two things were affected both by `liquidtype` at once). These are the relevant fields affected.

* `liquidtype`: Determines liquidtype for liquid flowing behavior (not new), but (NEW!) does no longer affect player movement (default: `"none"`, as before)
* `liquid_move_physics`: Here you can explicitly enable or disable liquidlike-movement physics (swim up/down, slowly sink, etc.) for the node, regardless of `liquidtype`. (default: enabled if `liquidtype ~= "none"` and disabled otherwise)

## To do

Idea for future PR: `move_resistance` is currently an `u8`, but only 8 values (0-7) are actually used. For the future, support for the full range of 0-255 could be added to make it much more flexible. But this might be too much for this PR.

## How to test

Give yourself all those liquid test nodes in DevTest. In particular, get these nodes out of the Chest of Everything (if you don't have the chest: `/stuff`):

* Move-resistant node
* Move-resistant node, liquidlike
* Climbable Move-resistant node
* No-swim Liquid Source Node

Play around with the nodes. Check if everything behaves as expected and if there is no regression of the "old" liquid nodes. IMPORTANT: All old worlds should behave exactly the same, as this PR must be backwards-compatible.

Here are some illustrations to help you better understand the nodes:
* "Move resistant node" is basically a cobweb. Movement through this node is slowed in all directions, including falling.
* "Climbable Move resistant node" is as if you combine a ladder with a cobweb.
* "Move-resistant node, liquidlike" is like water when you wade through it, but it doesn't flow or spread. There are no "Source" or "Flowing" variants.
* "No-swim Liquid Source/Flowing Node" is like air when you move through it (no swimming), but it flows like water.

Also, you might want to change this setting: `movement_liquid_fluidity`. Test once with default value and one with changed value. The movement speed should always be the same for "Move-resistant node" regardless of setting, but it should change with the setting value for "Move-resistant node, liquidlike".

## Use cases
* Cobwebs
* Thick foliage
* Force fields
* Liquid oxygen (flows like water but you fall and walk through it like air)

LoneWolfHT commented:
Does this work when you have older clients joining a new server?

Wuzzy2 commented:
Idk, I haven't tested. I don't know if this even matters. Will 5.4.0 come with a protocol bump? In that case, it doesn't matter anyway.

appgurueu commented:
@LoneWolfHT this feature will require both server and client to be 5.4.

hecktest commented:
Does this defer to viscosity to determine swimming resistance if the latter is nil? If not, this is potentially breaking.

v-rob commented:
> Does this defer to viscosity to determine swimming resistance if the latter is nil? If not, this is potentially breaking.

> If this field was not set, it will be equal to liquid_viscosity (for backwards-compability).

Zughy commented:
Any chance to include #7785 adding negative values to `move_resistance` ?

Wuzzy2 commented:
No. `move_resistance` is a slowdown modifier for moving though a non-`walkable` node. #7785 is concerned about the walking speed when walking on top of a `walkable` node. So these are 2 completely different things.

appgurueu commented:
> No. `move_resistance` is a slowdown modifier for moving though a non-`walkable` node. #7785 is concerned about the walking speed when walking on top of a `walkable` node. So these are 2 completely different things.

Reverse `move_resistance` could be useful though (but I can't think of a use case).

rubenwardy commented:
See #8613

Wuzzy2 commented:
Update! I have added `move_resistance` support for `climbable` nodes.

(I will squash the commits on request.)

appgurueu commented on games/devtest/mods/testnodes/liquids.lua:
```diff
@@ -80,4 +82,53 @@ for d=0, 8 do
+			{name = "testnodes_liquidsource_r"..d..".png"..mod, backface_culling = false},
+			{name = "testnodes_liquidsource_r"..d..".png"..mod, backface_culling = true},
+		},
+		alpha = 192,
```
Deprecated

appgurueu commented on src/nodedef.cpp:
```diff
@@ -618,6 +621,11 @@ void ContentFeatures::deSerialize(std::istream &is)
+		u8 tmp_move_resistance = readU8(is);
+		if (is.eof())
+			throw SerializationError("");
+		move_resistance = tmp_move_resistance;
```
What happens if a new client (5.5) connects to an old server (5.4) - the server would not send `move_resistance`, right? Wouldn't `move_resistance` of liquids be `0` then?

Wuzzy2 commented:
> What happens if a new client (5.5) connects to an old server (5.4)

As far I know, this is not permitted anyway. As far I know, Minetest does not seek compability for new clients to connect to old servers. (Correct me if I'm wrong.)

appgurueu commented:
> > What happens if a new client (5.5) connects to an old server (5.4)
> 
This is permitted, unless a setting is set.

appgurueu commented on src/nodedef.cpp:
```diff
@@ -618,6 +621,11 @@ void ContentFeatures::deSerialize(std::istream &is)
+
+		u8 tmp_move_resistance = readU8(is);
+		if (is.eof())
+			throw SerializationError("");
```
What about `move_resistance = liquid_viscosity` instead of `throw SerializationError("");`? That should fix the comment below.

Wuzzy2 commented:
@appgurueu: idk.

SmallJoker commented on src/client/localplayer.cpp:
```diff
@@ -228,6 +228,7 @@ void LocalPlayer::move(f32 dtime, Environment *env, f32 pos_max_d,
 		node = map->getNode(pp, &is_valid_position);
 		if (is_valid_position) {
 			in_liquid = nodemgr->get(node.getContent()).isLiquid();
+			move_resistance = nodemgr->get(node.getContent()).move_resistance;
 			liquid_viscosity = nodemgr->get(node.getContent()).liquid_viscosity;
```
```C++
const ContentFeatures &cf = nodemgr->get(node.getContent());
in_liquid = cf.isLiquid();
move_resistance = cf.move_resistance;
liquid_viscosity = cf.liquid_viscosity;
```
same for the other changes below. Alternatively it would be possible to use a `const ContentFeatures *liquid;` member for LocalPlayer to not copy this information all over again. That however would require more code changes.

SmallJoker commented on src/client/clientenvironment.cpp:
```diff
@@ -194,32 +194,43 @@ void ClientEnvironment::step(float dtime)
+					if (dl > lplayer->movement_liquid_fluidity_smooth)
+						dl = lplayer->movement_liquid_fluidity_smooth;
+				} else if (dl > (0.5f * BS)) {
+					dl = (0.5f * BS);
```
Why 0.5? Would you please be so nice to add a comment here?

appgurueu replied:
Also, please remove the unnecessary parentheses.

SmallJoker commented on src/script/common/c_content.cpp:
```diff
@@ -824,6 +827,10 @@ void read_content_features(lua_State *L, ContentFeatures &f, int index)
 
+	// How much the node slows down players, ranging from 1 to 7,
+	// the higher, the slower.
+	f.move_resistance = getintfield_default(L, index,
+			"move_resistance", f.move_resistance);
```
Value needs clamping, especially to zero before assigning to a `u8` variable.

appgurueu replied:
Clamping? I'd much prefer an error be thrown if a modder tries to set this to an invalid value.

SmallJoker commented on doc/lua_api.txt:
```diff
@@ -7529,6 +7529,10 @@ Used by `minetest.register_node`.
 
+        move_resistance = 0,
+        -- Slows down movement of players through this node (max. 7).
+        -- If this is nil, it will be equal to liquid_viscosity.
```
Sounds a bit weird.
```suggestion
        -- If nil or not specified, the value will default to the "liquid_viscosity".
```

appgurueu replied:
Sounds absolutely correct to a Lua developer. Unspecified and nil are pretty much the same thing (except for varargs, but that's largely irrelevant, this is about tables).

Wuzzy2 commented:
It turned out the 0.5 thing was not neccessary at all.
But I just noticed a conceptual problem with the PR. The decoupling of movement physics from the liquidtype was not actually complete. The movement code itself still could lead to different movement speeds and physics solely depending on the liquidtype. In other words, the liquidtype still had side-effect, so that was not a true decoupling.

I am going to add another field `has_liquidlike_physics` to allow to explicitly turn on/off liquid movement physics. This should complete the decoupling.

Wuzzy2 commented:
Good news! I have reworked the liquid movement code to completely decouple the player movement physics from `liquidtype`. With "liquid movement physics" I mean stuff like the ability of players to "swim" up/down in the node or that you sink slowly if not doing anything.
Previously, the physics were strongly coupled to `liquidtype`. I.e. if `liquidtype~="none", this not only enabled liquid flowing/spreading, but also changed the player movement physics. In contrast, if `liquidtype=="none"`, you can't have liquid movement physics.
This was kinda bad design IMO because liquidtype does two different things (node spreading, player movement) at once.

To do the decoupling, I added `liquid_move_physics`, a new node field allowing you to enable and disable liquid.

The following things are now possible:

* Nodes that flow/spread like liquids but players move through like in air (no swimming)
* Nodes that have liquid move physics with `liquidtype == "none"`, i.e. you can ignore setting all the other `liquid` node fields. In these nodes, you can swim up/down and sink like in a normal liquid, but you don't have to deal with all this spreading stuff
* Nodes that have a move resistance but no liquid physics. This is not new of the commit, I just mention it for completeness. In these nodes, you will move slower, it's very similar to how liquids work, but with a crucial difference: The special liquid movement rules do not apply: There is no up/down swimming, no liquid-style sinking, etc. That use case is great for cobwebs
* "climbable" liquids. That is, a node that spreads like a liquid, has liquid physics disabled but has climbable like a ladder
. This is achieved with `liquidtype="source"/"flowing"`, `liquid_move_physics="normal"` and `climbable=true`. True, this is a rather esoteric use case, and I actually didn't actively wrote code for it. This is just a logical consequence of the decoupling, which is nice, I guess. :-)

As for the move resistance, a few words: There are now 2 ways that move resistance works: If the node has liquid move physics disabled, the movement speed is always a fixed modifier, ignoring the `movement_liquid_*` settings. This is a change from the previous state of the PR, because it took the settings into account. This bothered me, because I realized that the `movement_liquid_*` setting would affect ALL `move_resistance` nodes, even those that were designed to be not a liquid. That would be an annoying side-effect for modders which also lead me to the idea of decoupling things even more.

The other way is when liquid move physics are enabled. In that case, the `movement_liquid_*` settings DO apply, like they already did for all nodes with `liquidtype~="none"`. You can test and compare this with the new move resistance nodes in DevTest (mave sure to modify `movement_liquid_fluidity` before to test the difference).

I believe all changes are backwards-compatible. Existing games and mods should not break.

New test nodes:
* Move resistance Node, liquidlike (like the previous move-resistant nodes, but you can also "swim" in them
* No-swim liquid

Wuzzy2 commented:
OK, I realize the previous post was very complicated, so here's the summary of it:

* The situation: `liquidtype` was strongly coupling liquid flowing physics (the spawning an spreading of flowing liquid nodes) with player liquid movement physics. You only could either have both, or neither of those things.
* I considered this to be bad design, so I decoupled it:
    * New node field `liquid_move_physics` added, this allows to explicitly enable or disable liquid move physics for ANY node
    * Old node field `liquidtype` now only directly affects flowing physics but no longer player liquid movement physics
* This opens up a few new interesting use cases
* It's backwards-compatible (sane default behavior)
* Read `lua_api.txt` for details

Wuzzy2 commented:
@SmallJoker: About the clamping of `move_resistance`: I disagree with this proposal. Reason: This function doesn't apply clamping to any other value so far.
I feel like such a change would make more sense if clamping/error checking is applied to ALL values at once, instead of adding a single odd check while the rest of the code is ignored. Maybe that's something for a different PR.

On the other hand, error checks/asserts might be much better anyway. If the modder provides a bad value (like -4359234), clamping seems like the completely wrong thing to do, we'd basically just hide an error.

Wuzzy2 commented:
We had a long discussion in IRC about compability concerns. We have determined that there will be inconsistencies if and old client (5.4.0 or earlier) connects to a new one. Specifically:
* Old clients will move through `move_resistance` nodes (cobwebs) like air
* But in no-swim liquids, old clients will move through like a liquid (but they should move through them like air)

For example, if a server has a map in which there's a cobweb trap, new clients will "sink" gracefully through it while old client fall through it like air, potentially receiving deadly fall damage. This is therefore a serious risk of massive frustration for the player base. This is not good.

There is no good fallback behavior I can think of. For example, a `move_resistance` in a non-liquid cannot be perfectly emulated in new clients.

There are several courses of actions regarding compability:

A) Try the best to emulate the movement physics of the new server. This requires to send various faked information to old clients to approximate the move behavior. Unfortunately, you also need to fake `liquidtype` for this, leading to various subtle but annoying side-effects, like with `liquids_pointable` tools being able to point a node on one client but not on another. This is a pretty messy and hacky solution
B) Do nothing. Leave the responsibiltiy to ensure correctness to server operators entirely. If a server operator needs the new liquid movement behavior, they could use `strict_protocol_version_checking` to make sure none of the awkward situations above can happen
C) Declare the next version to be incompatible to 5.4.0
D) Postpone this PR until 6.0.0

sfan5 commented on src/client/localplayer.cpp:
```diff
@@ -800,7 +804,8 @@ void LocalPlayer::old_move(f32 dtime, Environment *env, f32 pos_max_d,
-			in_liquid = nodemgr->get(node.getContent()).isLiquid();
+			in_liquid = nodemgr->get(node.getContent()).hasLiquidMovePhysics();
+			move_resistance = nodemgr->get(node.getContent()).move_resistance;
 			liquid_viscosity = nodemgr->get(node.getContent()).liquid_viscosity;
```
should be changed to call `nodemgr->get(node.getContent())` only once

sfan5 commented on src/client/localplayer.cpp:
```diff
@@ -810,7 +815,8 @@ void LocalPlayer::old_move(f32 dtime, Environment *env, f32 pos_max_d,
-			in_liquid = nodemgr->get(node.getContent()).isLiquid();
+			in_liquid = nodemgr->get(node.getContent()).hasLiquidMovePhysics();
+			move_resistance = nodemgr->get(node.getContent()).move_resistance;
 			liquid_viscosity = nodemgr->get(node.getContent()).liquid_viscosity;
```
same here

sfan5 commented on src/client/localplayer.h:
```diff
@@ -55,7 +55,9 @@ class LocalPlayer : public Player
+	// Slows down the player when moving through
+	u8 move_resistance = 0;
+	// Liquid flow slowness
 	u8 liquid_viscosity = 0;
```
Isn't this variable unused then?

sfan5 commented on src/nodedef.cpp:
```diff
@@ -618,6 +623,18 @@ void ContentFeatures::deSerialize(std::istream &is)
+
+		tmp = readU8(is);
+		if (is.eof()) {
+			move_resistance = liquid_viscosity;
```
The `move_resistance = liquid_viscosity;` should happen further up instead of mixing this into the serialization code here IMO

sfan5 commented on src/nodedef.cpp:
```diff
@@ -618,6 +623,18 @@ void ContentFeatures::deSerialize(std::istream &is)
+		}
+		move_resistance = tmp;
+
+		enum LiquidMoveType tmp_lmt = (enum LiquidMoveType) readU8(is);
+		if (is.eof())
+			throw SerializationError("");
+		liquid_move_physics = tmp_lmt;
```
```suggestion
		tmp = readU8(is);
		if (is.eof())
			throw SerializationError("");
		liquid_move_physics = static_cast<enum LiquidMoveType>(tmp);
```

sfan5 commented on src/nodedef.h:
```diff
@@ -469,6 +480,11 @@ struct ContentFeatures
+	bool hasLiquidMovePhysics() const{
+		return (liquid_move_physics == LIQUID_MOVE_LIQUID) ||
+			(liquid_move_physics == LIQUID_MOVE_AUTO &&
+			liquid_type != LIQUID_NONE);
```
Wouldn't it make more sense to do the auto -> liquid or not decision in `read_content_features()`?

Wuzzy2 commented:
> Defaults are usually nil so why "auto" here?

It's analogous to `liquidtype` which is also an enum type with "none", "source" and "flowing".

> I think "none" would be better here
> or actually, can't this be a boolean?

It's internally an enum like `liquidtype` becaue there are 3 possible states, not 2. The `"auto"` state is required for backwards-compability (liquid physics are derived from `liquidtype`).

> Isn't this variable unused then?

Hmmm. Technically, yes. The client modding API needs to be taken care of somehow. It currently returns `liquid_viscosity` to the client but this actually means the physical move resistance, which is also what this is used for (ignoring liquid flow speed, which is server-side). I should probably redo documentation or something ...

> Wouldn't it make more sense to do the auto -> liquid or not decision in read_content_features()?

Not really. Because even then I would require the function, unless I explicitly repeat the check every time.

Wuzzy2 commented:
So, here's a question: What should I do with `get_liquid_viscosity()` of the client API? Should I interpret it to return `move_resistance`, since that was the original intention of this function? Or do you have a better idea?

sfan5 commented:
You missed my suggestion to rename `"normal"` to `"none"` for `liquid_move_physics`. Seeing the combination "normal liquid physics" definitely does not make the think that there are *no* liquid physics.

> Not really. Because even then I would require the function, unless I explicitly repeat the check every time.

I know, this suggestion was not so you could remove the function. Turning it into a bool would remove the need for an enum entirely, both in the mod API and internally.

> What should I do with get_liquid_viscosity() of the client API?

You could technically just remove it, the CSM API is still declared experimental.

Wuzzy2 commented:
Okay, I have completed the remaining tasks:

* `liquid_move_physics` is now a boolean with `nil` for automatic
* Removed `get_liquid_viscosity()` from client Lua

Wuzzy2 commented:
I noticed due to my changes, the term "liquid node" became a little ambigious in the `lua_api.txt` so I clarified what exactly the term "liquid node" means in the documentation where it is relevant. This was important because "liquid node" is not that clear. Does it mean liquidtype? The drawtype? Or nodes with liquid player movement physics enabled? (the correct answer is "liquidtype" but the new developers don't know that).

sfan5 commented on src/nodedef.cpp:
```diff
@@ -618,6 +624,17 @@ void ContentFeatures::deSerialize(std::istream &is)
 		alpha = static_cast<enum AlphaMode>(tmp);
+
+		tmp = readU8(is);
+		if (is.eof()) {
+			throw SerializationError("");
+		}
```
```suggestion
		if (is.eof())
			throw SerializationError("");
```

sfan5 commented on src/nodedef.h:
```diff
@@ -90,6 +90,13 @@ enum LiquidType
 	LIQUID_SOURCE,
 };
 
+enum LiquidMoveType
+{
+	LIQUID_MOVE_AUTO,
+	LIQUID_MOVE_NONE,
+	LIQUID_MOVE_LIQUID,
+};
+
```
should be removed

sfan5 commented on src/script/common/c_content.cpp:
```diff
@@ -719,6 +719,9 @@ void read_content_features(lua_State *L, ContentFeatures &f, int index)
 	// the slowest possible
 	f.liquid_viscosity = getintfield_default(L, index,
 			"liquid_viscosity", f.liquid_viscosity);
+	// If move_resistance is not set explicitly,
+        // move_resistance is equal to liquid_viscosity
```
```suggestion
	// If move_resistance is not set explicitly, it is equal to liquid_viscosity
```

sfan5 commented on src/script/common/c_content.cpp:
```diff
@@ -830,6 +848,7 @@ void push_content_features(lua_State *L, const ContentFeatures &c)
 	std::string paramtype2(ScriptApiNode::es_ContentParamType2[(int)c.param_type_2].str);
 	std::string drawtype(ScriptApiNode::es_DrawType[(int)c.drawtype].str);
 	std::string liquid_type(ScriptApiNode::es_LiquidType[(int)c.liquid_type].str);
+	std::string liquid_move_physics(ScriptApiNode::es_LiquidMoveType[(int)c.liquid_move_physics].str);
```
This is wrong now that `liquid_move_physics` is a boolean.

sfan5 commented on src/script/cpp_api/s_node.cpp:
```diff
@@ -77,6 +77,14 @@ struct EnumString ScriptApiNode::es_LiquidType[] =
+		{LIQUID_MOVE_NONE, "none"},
+		{LIQUID_MOVE_LIQUID, "liquid"},
+		{0, NULL},
+	};
```
Needs to be removed (too)

sfan5 reviewed (commented):
Works but there are some things that need to be fixed

Wuzzy2 commented:
All done.

sfan5 commented on src/nodedef.cpp:
```diff
@@ -618,6 +624,16 @@ void ContentFeatures::deSerialize(std::istream &is)
+			throw SerializationError("");
+		liquid_move_physics = tmp;
 	} catch(SerializationError &e) {};
 }
```
Wait, you're missing code in this function to set `liquid_move_physics` to true if the server doesn't provide a value

Wuzzy2 commented:
Please try again.

NathanSalapat commented:
Sweet, will be nice to use this feature for bushes and thickets.

Wuzzy2 commented:
Follow-up PR: #11665
